src/anyio/_core/_contextmanagers.py

```diff
@@ -0,0 +1,136 @@
+from __future__ import annotations
+
+from abc import abstractmethod
+from collections.abc import AsyncGenerator, Generator
+from inspect import isasyncgen, iscoroutine
+from types import TracebackType
+from typing import Generic, TypeVar, final
+
+T = TypeVar("T")
+
+
+class ContextManagerMixin(Generic[T]):
+    """
+    Mixin class providing context manager functionality via a generator-based
+    implementation.
+
+    This class allows you to implement a context manager using a generator, similar to
+    :func:`contextlib.contextmanager`, but for classes that need to embed other context
+    managers.
+
+    This class is designed to streamline the use of context management by
+    requiring the implementation of the `__contextmanager__` method, which
+    should yield instances of the class itself. It then wraps this generator
```

Comment:
Well, there's no requirement for a context manager to return Self and the wrapper shouldn't enforce that. It should accept / forward whatever type `__contextmanager__` is declared to return.

On the other hand, what this wrapper does not do is to ensure that the context is not entered twice, and *that* **really** is a bug.

Reply:
> Well, there's no requirement for a context manager to return Self and the wrapper shouldn't enforce that. It should accept / forward whatever type __contextmanager__ is declared to return.

I've already updated the docstring before you posted this. This version was never intended to be pushed to GH.

> On the other hand, what this wrapper does not do is to ensure that the context is not entered twice, and that really is a bug.

Hm, I see what you mean. But rather than raise an error here, perhaps reentrancy should be properly supported instead?

Reply:
I've thus disallowed reentrancy.